pkg/pillar/cmd/nodeagent/nodeagent.go

```diff
@@ -0,0 +1,504 @@
+// Copyright (c) 2017-2018 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// NodeAgent interfaces with baseosmgr for baseos upgrade and test validation
+// we will transition through zboot for baseos upgrade validation process
+
+// nodeagent publishes the following topic
+//   * zboot config                 <nodeagent>  / <zboot> / <config>
+//   * nodeagent status             <nodeagent>  / <status>
+
+// nodeagent subscribes to the following topics
+//   * global config
+//   * base os status               <baseosmgr> / <baseos> / <status>
+//   * zboot status                 <baseosmgr> / <zboot> / <status>
+//   * ledblink config              <zedagent>  /<ledblink> / <config>
+//   * zedagent status              <zedagent>/ <status>
+
+package nodeagent
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/lf-edge/eve/pkg/pillar/agentlog"
+	"github.com/lf-edge/eve/pkg/pillar/cast"
+	"github.com/lf-edge/eve/pkg/pillar/pidfile"
+	"github.com/lf-edge/eve/pkg/pillar/pubsub"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/zboot"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	agentName             = "nodeagent"
+	maxConfigGetFailCount = 5
```

Comment:
How is this useful? I we configure the get timer to every 5 seconds, this means a mere 25 seconds.

Is there a reason the time limit doesn't work??

Reply:
I am seeing some getconfig request fails in between. To avoid the transient get request fail, I have added some default count, until we reset the upgrade validation test function, and start the test again.

Reply:
But you'll fail after 25 seconds AFAICT even if the timer is set to be a lot higher.

Reply:
The controller get config request fails, periodically.. has a pattern of some temporary failure, and the next  config get request suceeds. In the 25 seconds, it would have recovered, and we will complete the upgrade validation test. 

{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:42256-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:37.616136477Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:39846-\u003e192.168.0.1:53: i/o timeout] No management interfaces]","time":"2019-10-01T15:54:42.620100823Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:55826-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:42.629004753Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:35562-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:48.671905065Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:54735-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:48.680837667Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:59296-\u003e192.168.0.1:53: i/o timeout] No management interfaces]","time":"2019-10-01T15:54:55.255563556Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:53224-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:55.276441046Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:44054-\u003e192.168.0.1:53: i/o timeout] No management interfaces]","time":"2019-10-01T16:31:40.852548504Z"}

Reply:
That's a bug in the DNS handling in nim/devicenetwork which I just merged a fix into master.

But it isn't a count which matters; it is the 600 second timeout which matters. We shouldn't have any hardcoded numbers like this.

Reply:
Sure, have made the changes, and also removed the ledblinker config for deciding the device registration state.